designs/drift-hash-versioning.md

```diff
@@ -0,0 +1,180 @@
+
+Option 1 will have two case for drift:
+
+1. When the CRD `nodepool-hash-version` matches between the `NodePool` and `NodeClaim`:
+    1. Drift will work as normal and the `NodePool` hash will be compared to evaluate for drift 
+2. When the CRD `nodepool-hash-version` does not match between the `NodePool` and `NodeClaim`:
+    1. Karpenter would recalculate the hash of the `Nodepool` and back-propagate the annotation to the NodeClaims
+      - Karpenter will need not back-propagate the `nodepool-crd-hash` annotation `NodeClaims` that are considered drifted.
+    2. Drift will work as normal once a `NodePool` field is updated
+
+Pros/Cons
+
+- 👍👍 Simple implementation effort 
+- 👍 Cover general cases of updates to the `NodePool` CRD or the hashing mechanism
+- 👎 Nodes that were considered drifted before the Karpenter upgrade are not able to be un-drifted after a Karpenter upgrade 
+- 👎 Does not handle cases where users change `NodePool` resources as Karpenter is updating
+
+### Option 1B: Fallback to individual field evaluation for drift
+
+Similar to Option 1, there will be two cases for drift:
+
+1. When the `nodepool-hash-version` matches between the `NodePool` and `NodeClaim`:
+    1. Drift will work as normal and the `NodePool` hash will be compared to evaluate for drift 
+2. When the `nodepool-hash-version` does not match between the `NodePool` and `NodeClaim`:
+    1. Karpenter will individually check each field for drift. *This would be an adaptation of the (Deployment Controller)[https://github.com/kubernetes/kubernetes/blob/83e6636096548f4b49a94ece6bc4a26fd67e4449/pkg/controller/deployment/util/deployment_util.go#L603] for podTemplateSpec relationship to pods.*
+    2. Karpenter will need to populate the `NodeClaim` with values that were used in launching the instances. Drift evaluation will need a point of reference on manual checking.
+
+Pros/Cons
+
+- 👍 Cover general cases of updates to the `NodePool` CRD or the hashing mechanism 
+- 👍 Handles cases where users change `NodePool` resources as Karpenter is updating
+- 👎 Cloud-provider would need to propagate on to the NodeClaim to adopt this method for static drift checking 
+- 👎👎 Negatively impact on performance for Karpenter until all NodeClaims match the `nodepool-hash-version`
+- 👎👎 A high maintenance effort required in supporting individual field evaluation 
+
+### Option 2: Update the NodePool API Version and utilize Conversion webhooks to update the `nodepool-hash`    
+
+The Karpenter APIVersion of the CRDs will be changed each time their is a change as such: 
+
+1. A `NodePool` adds default values to existing fields that are included in the hash calculations
+2. A `NodePool` field is added to the hash calculation with an already-set value
+3. A `NodePool` field is removed from the hash calculations
+
+As part of retrieving the `NodePool` from the API Server, a converion webhook would recaluclate the hash and propagte the new hash to the `NodeClaims`.
```

Comment:
I don't think the conversion webhook would be able to re-propoagate it down to the NodeClaims, I think you mean that the conversion webhook would convert into the new hash scheme on the NodePool and NodeClasses. The other piece that I think is missing here is that you _still_ need a `karpenter.sh/nodepool-hash-version` annotation (or something similar to it) in this solution. Otherwise, how does the controller know that it needs to re-propagate the updated hash down to the NodeClaims after it converts?